torchx/lib/torchx/backend.ex

```diff
@@ -0,0 +1,219 @@
+  end
+
+  @impl true
+  def random_normal(%{type: _type, shape: shape} = out, mu, sigma) do
+    NIF.normal(mu, sigma, shape) |> from_ref(out)
+  end
+
+  ## Transfer
+
+  @impl true
+  def to_batched_list(%{shape: shape} = out, %{data: %{ref: ref}}),
+    do: NIF.split(ref, elem(shape, 0)) |> from_ref(out)
+
+  @big_tensor_threshold 10_000_000
+
+  def to_binary(_tensor, _limit \\ nil) do
+    raise "Operation to_binary is not supported on Torchx.Backend. " <>
+            "You must first transfer the tensor to Elixir by calling Nx.backend_transfer/1"
+  end
+
+  defp to_blob(%{type: {_, elem_size}, data: %{ref: ref}} = tensor, limit \\ nil) do
+    if(limit,
+      do: NIF.to_blob(ref, limit),
+      else: NIF.to_blob(ref)
+    )
+  end
+
+  @impl true
+  def backend_deallocate(%Nx.Tensor{data: %{ref: ref}}), do: NIF.delete_tensor(ref)
+
+  @impl true
+  def backend_transfer(tensor, Nx.Tensor, opts) do
+    backend_transfer(tensor, Nx.BinaryBackend, opts)
+  end
+
+  def backend_transfer(tensor, Torchx.Backend, _opts) do
+    tensor
+  end
+
+  def backend_transfer(tensor, backend, opts) do
+    backend.from_binary(tensor, to_blob(tensor), opts)
+  end
+
+  @impl true
+  def from_binary(%{type: type, shape: shape} = out, binary, _opts) do
+    NIF.from_blob(binary, shape, torch_type(type)) |> from_ref(out)
+  end
+
+  ## Shape
+
+  @impl true
+  def reshape(out, %{data: %{ref: ref}} = tensor, shape),
+    do: NIF.reshape(ref, shape) |> from_ref(tensor)
+
+  @impl true
+  def as_type(%{type: type} = out, %{data: %{ref: ref}} = tensor),
+    do: NIF.to_type(ref, torch_type(type)) |> from_ref(out)
+
+  @impl true
+  def squeeze(out, tensor, _axes) do
+    NIF.squeeze(tensor.data.ref) |> from_ref(out)
+  end
+
+  @impl true
+  def transpose(out, tensor, axes), do: IO.inspect(axes)
+  # do: NIF.transpose(tensor, dim0, dim1) |> from_ref(out)
+
+  ## Ops
+
+  @impl true
+  def add(out, left, right) do
+    NIF.add(left.data.ref, right.data.ref) |> from_ref(out)
+  end
+
+  @impl true
+  def dot(
+        out,
+        %{type: t1, data: %{ref: left_ref}} = left,
+        _axes1,
+        %{type: t2, data: %{ref: right_ref}} = right,
+        _axes2
+      ) do
+    NIF.dot(left_ref, right_ref) |> from_ref(out)
+  end
+
+  @impl true
+  def cholesky(%{type: output_type, shape: {rows, cols}} = out, tensor) do
+    NIF.cholesky(tensor.data.ref) |> from_ref(out)
+  end
+
+  @impl true
+  def qr(
+        {q_holder, r_holder},
+        tensor,
+        opts
+      ),
+      do: NIF.qr(tensor.data.ref, opts[:mode] == :reduced) |> from_ref({q_holder, r_holder})
+
+  @impl true
+  def inspect(%{type: {_, elem_size}} = tensor, inspect_opts) do
+    limit = if(inspect_opts.limit == :infinity, do: nil, else: inspect_opts.limit + 1)
+
+    if on_cpu?(tensor) do
+      byte_size = nbytes(tensor)
+      byte_limit = limit && limit * div(elem_size, 8)
+
+      if(min(byte_limit, byte_size) > @big_tensor_threshold) do
```

Comment:
Nitpick: no parens around if:

```suggestion
      if min(byte_limit, byte_size) > @big_tensor_threshold do
```